parrot/src/pfs_dispatch.cc

```diff
@@ -973,6 +977,11 @@ static void decode_execve( struct pfs_process *p, int entering, INT64_T syscall,
 failure:
 		divert_to_dummy(p, -errno);
 done:
+		p->euid = new_uid;
+		p->suid = new_uid;
+		p->egid = new_gid;
+		p->sgid = new_gid;
```

Comment:
These uid/gid changes need to be made only if execve succeeds. Otherwise, a failed execve of a setuid executable would raise privileges even though the setuid executable is not running.

Fixing this require saving the changes in the pfs_process struct. The `execve` success is here:

````
      if (actual == 0) {
            p->table->complete_at_path(AT_FDCWD, p->new_logical_name, p->name);
            debug(D_PROCESS, "execve: %s (%s) succeeded in 64-bit mode", p->new_logical_name, p->name);
````